helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -127,7 +127,9 @@ public class BaseControllerDataProvider implements ControlContextProvider {
   private final Set<String> _enabledLiveSwapInInstanceNames = new HashSet<>();
   private final Map<String, MonitoredAbnormalResolver> _abnormalStateResolverMap = new HashMap<>();
   private final Set<String> _timedOutInstanceDuringMaintenance = new HashSet<>();
-  private Map<String, LiveInstance> _liveInstanceExcludeTimedOutForMaintenance = new HashMap<>();
+  private Map<String, LiveInstance> _allLiveInstanceExcludeTimedOutForMaintenance = new HashMap<>();
+  private Map<String, LiveInstance> _assignableLiveInstanceExcludeTimedOutForMaintenance =
+      new HashMap<>();
```

Comment:
We maintain separate maps because assignable is a subset. We have many calls to getAssignableLiveInstances and would have to rebuild this subset Map each time if we don't maintain another and rebuild it only when dataRefresh happens.